rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+        // Here we have a gap signaled by the input being null, but a greater
+        // than 0 input_len which provides the size of the gap.
+        state_safe.on_request_gap(input_len);
+        AppLayerResult::ok()
+    } else {
+        let buf: &[u8];
+        buf = build_slice!(input, input_len as usize);
+        state_safe.parse_request(buf)
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_parse_response(
+    flow: *const Flow, state: *mut std::os::raw::c_void, pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    let _eof = if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF_TC) > 0 {
+        true
+    } else {
+        false
+    };
+    let state_safe: &mut PgsqlState;
+    state_safe = cast_pointer!(state, PgsqlState);
+
+    if input == std::ptr::null_mut() && input_len > 0 {
+        // Here we have a gap signaled by the input being null, but a greater
+        // than 0 input_len which provides the size of the gap.
+        state_safe.on_response_gap(input_len);
+        AppLayerResult::ok()
+    } else {
+        let buf: &[u8];
+        buf = build_slice!(input, input_len as usize);
+        state_safe.parse_response(buf, flow)
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_state_get_tx(
+    state: *mut std::os::raw::c_void, tx_id: u64,
+) -> *mut std::os::raw::c_void {
+    let state_safe: &mut PgsqlState = cast_pointer!(state, PgsqlState);
+    match state_safe.get_tx(tx_id) {
+        Some(tx) => {
+            return tx as *const _ as *mut _;
+        }
+        None => {
+            return std::ptr::null_mut();
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_tx_count(state: *mut std::os::raw::c_void) -> u64 {
+    let state_safe: &mut PgsqlState;
+    unsafe {
+        state_safe = cast_pointer!(state, PgsqlState);
+    }
+    return state_safe.tx_id;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_tx_get_state(tx: *mut std::os::raw::c_void) -> PgsqlTransactionState {
+    let tx_safe: &mut PgsqlTransaction;
+    unsafe {
+        tx_safe = cast_pointer!(tx, PgsqlTransaction);
+    }
+    return tx_safe.tx_state;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_tx_get_alstate_progress(
+    tx: *mut std::os::raw::c_void, _direction: u8,
+) -> std::os::raw::c_int {
+    return rs_pgsql_tx_get_state(tx) as i32;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_events(
+    tx: *mut std::os::raw::c_void,
+) -> *mut core::AppLayerDecoderEvents {
+    let tx_safe: &mut PgsqlTransaction;
+    unsafe {
+        tx_safe = cast_pointer!(tx, PgsqlTransaction);
+    }
+    return tx_safe.events;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_event_info(
+    _event_name: *const std::os::raw::c_char, _event_id: *mut std::os::raw::c_int,
+    _event_type: *mut core::AppLayerEventType,
+) -> std::os::raw::c_int {
+    return -1;
```

Comment:
Since this is an optional function, if we are not doing anything here, maybe we can set it as `None` in the registration table.

Reply:
Makes sense. I was wondering if I should get rid of the functions that came with the template but are not used _yet_...